### PR TITLE
Govern slice expressions by slicing domain (and enable sparse slicing)

### LINKED ISSUE #12272: Design musings: Relationship between a slice expression and its domain
I've been wrestling recently with making slices of distributed arrays more efficient, and am getting caught up in some questions about the relationship between a slice expression and its domain (either in the sense of "the domain specifying the slice" and/or "the domain of the slice expression itself").

In more detail, there are a few choices for what to do with the domain specified in an array slice, `A[D]`:

1) have the slice refer to the original domain, `D`
2) make a copy of `D` and have the slice refer to that
3) create a new domain that snapshots `D`'s indices as a default rectangular domain
4) create a new domain that snapshots `D`'s indices using `A`'s domain map / distribution
5) something else

Currently, master takes approach 4, but this has some downsides.  Here's a rundown of each approach:

1) have the slice refer to the original domain, `D`

    In many respects, this seems natural / intuitive / cheap to implement, but it has some drawbacks:
    - **subject to domain changes:** if `D` is re-assigned while a reference to the slice still exists, the slice will describe different indices; this might be acceptable with sufficient documentation / notice, but also seems somewhat surprising or unintuitive.  For example:
      ```chapel
      const D = {1..10};
      ref AD = A[D];
      D = {1..20};
      writeln(AD);       // should 10 or 20 elements be printed?
      ```
    - **non-distributed slices of distributed arrays:** if `A` is distributed and `D` is not (say it's just an anonymous domain), then the slice's domain is intuitively local.  For example, imagine:
      ```chapel
      const D = {1..n, 1..n} dmapped Block(...);
      var A: [D] real;
      ref AInner = A[{2..n-1, 2..n-1}];   // Or equivalently, A[2..n-1, 2..n-1].
      forall ij in AInner.domain do       // Presumably one would want this distributed,
        A[ij] = ...;                      // but naively, it wouldn't be.
      var B: [AInner.domain] real;        // Ditto.

   **THE FOLLOWING MAY BE INCORRECT**

    - **change in slicing semantics:** the current slicing semantics are that we intersect `D` with the domain of `A`, permitting a strided array to be sliced with a non-strided domain or an unbounded range.  Thus, the natural domain of a slice expression is not merely `D` but rather the intersection of `D` and `A.domain`.  Or we need to change the current slicing semantics.  For this reason all references to "make a copy of `D`" below should really be saying "make a copy of `D` intersected with `A.domain` (unless we change).

2) make a copy of `D` and have the slice refer to that

    This is similar to option 1, but has the advantage of not making the slice subject to changes to `D` that might occur during its lifetime (i.e., it eliminates the "subject to domain changes").  It still has the "non-distributed slices of distributed arrays" concern and adds another:
    - **distributed domain creation:** if `D` is a block-distributed domain, it has the downside of creating a new block-distributed domain, which is currently an expensive operation, involving all locales, and an arguably unnecessary one given that `D` was already distributed.


3) create a new domain that snapshots `D`'s indices as a default rectangular domain

    The idea here is to snapshot `D`'s indices to eliminate the "subject to domain changes" issue, but to use a default rectangular domain because they're cheap to create (eliminates the "distributed domain creation" concern of number 2.  However, it still has the "non-distributed slices of distributed arrays" concern and applies it to all cases.

4) create a new domain that snapshots `D`'s indices using `A`'s domain map / distribution

    This is what master currently does, and it has the advantage of eliminating the "subject to domain changes" issue and the "non-distributed slices of distributed arrays" concern, but incurs the "distributed domain creation" overheads for all slicing domains (local and distributed) and also has the following downside:
    - **small slices are expensive** This was one of the motivating cases for this work which is that a slice like `A[i..i+1, j..j+1]` will create a globally distributed block domain, involving every locale in its creation, even if the 4 elements being referenced are completely local to a single locale, perhaps even the current one.  It seems a shame to talk to everyone if it's not necessary.

5) something else
    Here are some other ideas I've been playing around with:
    a) use D if its distribution matches A's and copy it using A's distribution otherwise
      - this has the "subject to domain changes" and "small slices are expensive" downsides
    
    b) use D if it is const and its distribution matches A's, and copy it using A's distribution otherwise
      - this just has the "small slices are expensive" downside

      c) create a snapshot of D as a default rectangular domain and address the "non-distributed slices of distributed arrays" problems by creating the distributed domain lazily if needed / specializing operations to do the right thing in the distributed context.
      - this seems pretty complex / fragile to implement, but potentially addresses all of the above concerns
    
      d) create a more general way of lazily propagating definitions of distributed / privatized objects.  Today, when we create a distributed domain / array, we aggressively privatize it by telling every locale about it and how to make a local copy of the structure.  What if, instead, we had a way of creating it locally, giving it a pid, and then having other locales create their privatized copy of the data structure only when they're given a pid that they're not familiar with?  This laziness would typically not fire for arrays (since they tend to need their O(n) storage to be allocated and initialized), nor domains used to declare arrays (since they'd need to exist on the locale for the array to be created... though maybe they could be set up there only when the array is declared rather than as a separate step); but for domains like the ones used for slicing, it could permit them to be set up locally and then only communicated to locales involved in operations over them.  I think that downsides to this approach are that it is complex and may result in more (or more expensive) communication for data structures that ultimately need to be communicated to all locales (because if I look up a pid and realize I don't have it, I have to communicate back to you to learn what I need to to create my own copy).  But I think that the upsides might be worth it.
    
    e) A variant on (d) in which a domain map could specify whether a domain/array should be privatized lazily or proactively.  Perhaps full-fledged user variables would always be proactive (as they are today) but that domains used in slicing (say) would be lazy?

### LINKED ISSUE #12289: Are array slices snapshots or ongoing definitions?
This issue asks whether slices should be interpreted as an instantaneous snapshot of a sub-array expression or an array expression that specifies an ongoing relationship with respect to its domain and array.  I don't believe that this question comes up in the common case, yet is important to have an answer for it in the general case.

Consider the following code:

```chapel
var D = {1..n};   // declare a domain and array
var A: [D] real;

var Inner = {2..n-1};
ref AS = A[Inner];  // capture a reference to A sliced by Inner

Inner = {3..n-2};   // modify inner
writeln(AS);      // refer to the slice
```

Should the code above print `A[2..n-1]` (the definition of the slice at the time the reference to it was created) or `A[3..n-2]` (the definition of the slice based on the revised definition of `Inner`)?  The argument in favor of the former is that it is sufficiently intuitive that it's likely what most new Chapel programmers would expect (though, again, I think it's most likely that they may never write code that runs into this situation).  The argument in favor of the latter is that slices are array expressions, all array expressions in Chapel have a domain and an ongoing relationship with that domain, and `Inner` is a natural domain for the slice's virtual array.  One other argument in favor of it is that the first interpretation requires making a copy of `Inner` in the event that it's not constant and such copies can be expensive in the event that `Inner` is a distributed domain.

Here's a second, and slightly more esoteric, example using the same declarations as above:

```chapel
ref AS2 = A[4..];  // capture a reference to a slice of A
D = {1..2*n};       // resize A

writeln(AS2);   // refer to the slice
```

Should this second example print out `A[4..n]` or `A[4..2*n]`?

Note that this decision can arguably be made independently of the previous one because it comes down to whether the anonymous domain that we create for the unbounded range `4..` in the slice expression is (a) `{4..}` (a domain with an unbounded range as proposed in issue #10595) or (b) `{4..n}` (we take the intersection with the array's domain at the time of the slice's creation to create a bounded domain at that point in time).  And since the anonymous domain can't be re-assigned, the original question doesn't come into play.  That said, the two questions share a thematic similarity.

### LINKED ISSUE #12704: Add support for slicing a dense array using a sparse slice
As a Chapel user, I'd like to slice dense arrays using sparse domains because it's a natural and compact way of selecting a subset of array elements.  It's also a composition of features that's been designed as intended for use since Chapel's inception, yet never supported.

### LINKED ISSUE #12873: Make slice expressions using domains be governed by that domain
As a Chapel programmer, when I slice an array by a domain, I would like the slice expression governed by that domain in order to avoid unnecessary copies, reflect dynamic modifications to the domain, and mirror Chapel semantics for other array declarations and expressions.

This issue is a follow-on to https://github.com/Cray/chapel-private/issues/95 and https://github.com/Cray/chapel-private/issues/66 and is a stepping stone toward #12704 and towards reduce communication for certain slicing cases (https://github.com/chapel-lang/chapel/issues/9593).

### PR DESCRIPTION
**Pull request by bradcray**
Overview
=======
This PR changes the semantics of array slice expressions such that they no longer make a personal copy of the slicing domain, but instead refer to the slicing domain and are governed by it from that point forward.  "governed" means things like:

* the domain of the slice expression is the slicing domain, such that changes to the domain are reflected in the slice expression if it is still alive (e.g., see new test sliceUsesDom.chpl)
* new arrays created in terms of a slice expression will also use that domain as their defining domain (e.g., see sliceUsesDom2.chpl)
* iterations over the slice expression respect the locality of the slicing domain (e.g., see sliceUsesDom-locality.chpl and note that we may change this in the future, see "NOTE ON LOCALITY" below)

As a side-effect of this work, slicing dense arrays via sparse domains is supported for the first time (after being demonstrated for over a decade in my powerpoint slides...).

Why?
=====
The reasons for this change have been wrestled with on issues #12873, #12289, and #12272, but the gist of the idea is:

* that the result is lower overhead, particularly for expensive domain formats like distributed domains, associative domains, sparse domains
* that the result is more similar to existing relationships between arrays and their domains
* that supporting these semantics provides an easy path for users who want the snapshot semantics (they can copy the domain), but that the old semantics made it difficult for a user to get these semantics.

How?
=====
The approach taken here is fairly straightforward:  Where before we took the slicing domain, unwrapped it and passed it to the slice-by-ranges routine, now we just associate the slice array expression with the original domain.  This permits us to slice by domains that do not have ranges (such as sparse domains) for the first time (e.g., see new tests sparseSliceDense*.chpl).

Note that associating the slice array expression with the slicing domain is a bit interesting because if we literally add it to the list of arrays in the domain (as we would for most arrays), we try to reallocate it when the domain is re-assigned as we would a normal array.  However, slices don't need to be reallocated (since the underlying array isn't governed by the slice and the slicing domain is already referenced in-place).  And moreover, they fail in this code path today because we try to dynamically cast each of the arrays in the list to some type (like BaseArr) before calling dsiReallocate on them, and this fails for SliceArrayViews (this could potentially be addressed, but there's no real motivation to do so since we don't want to resize them anyway and this is strictly a cheaper way to track domain lifetime: by bumping an integer as we'd also like to do for arrays declared over constant domains).  

File-by-file breakdown
================

Core changes
--------------
* ChapelArray.chpl: updates the slice-by-domain routine to use the existing domain; adds a bounds-checking helper for domains; updates the calls to add and remove arrays to avoid interacting with the domain's list of arrays
* ChapelDistribution.chpl: updates the code paths that add/remove domains from the list to only do it if the callee requests it (i.e., in all non-slice cases)

New tests
----------
* sliceKeepsDomainAlive.*: new test ensuring that a slice which outlives its domain is sufficient to keep that domain alive
* sliceUsesDom-locality.*: new test showing that locality for a slice follows the domain, not the array
* sliceUsesDom.*: new test showing that changing a domain will change the view of a slice that persists across the change
* sliceUsesDom2.*: new test showing that new expressions created in terms of a slice expression inherit the same domain
* sparseSliceDense*: demonstrates that we can now slice dense arrays by sparse ones

Performance Improvements
----------------------------
* assignReindex*.good: Communication decreases due to not needing to make copies of distributed domains when slicing

Test Updates
--------------
* localReads.*: Stencil domains permit slightly OOB accesses, but this test was using DR slices.  Update it to use explicit Stencil domains for the slices (to preserve the old behavior) and to slice with the old DR domains to show the OOB behavior (approved by @benharsh)
* testReplicatedDist.*: Like localReads.chpl, the behavior of this test was sensitive to whether the slicing domains were Block or DR.  Here, I cloned many of the interesting cases to show that slices are now sensitive to this choice (approved by @vasslitvinov)
* optimizations/bulkcomm/alberto/*: Like the above, these were slicing by a DR domain; I changed them to an explicit block-distributed domain to preserve the old behavior

Note on Locality
============
This PR necessarily takes a stance on a challenging semantic question about whether iterating over A[D] should follow A's distribution rather than D's (say A is distributed and D is not), and we may want to revisit this as we get more experience with it.  The reasons for taking the current approach of following D's distribution include:

* it gives users more direct control over where computation over their arrays takes place (say you want to take a small slice of a distributed array and have the result be local, this gives you those semantics
* it avoids having to move expensive domains (say, distributed sparse or associative domains) around between locales but uses them in-place; of course the dual of this is that it may move the array elements accessed around

I've opened up issue #12936 to capture some more thoughts around the choice of interpretation of domain-based slicing and foster additional discussion.

But there may be a learning curve here, so we might imagine having the support for (the ability to opt into?) warnings when slicing a distributed array using a domain that is local and spans multiple locales, or is distributed differently, or ...?

Resolved Issues
============

Resolves #12873.
Resolves #12289.
Resolves #12272.
Resolves #12704.

### CODE REVIEW

File: test/distributions/vass/testReplicatedDist1.chpl [hunk 1]
```diff
@@ -34,7 +34,7 @@ proc reset() {
   // explicitly go to each locale
   coforall loc in repllocales do on loc do
     // explicitly index into each element
-    for ix in Dsub do ARepl[ix] = 100 + here.id;
+    for ix in DsubLoc do ARepl[ix] = 100 + here.id;
 }
 
 /*
```
Comment (vasslitvinov) [inline test/distributions/vass/testReplicatedDist1.chpl]: Let's change Dsub -> DsubLoc in the (commented out) code below.

File: test/distributions/vass/testReplicatedDist1.chpl [hunk 2]
```diff
@@ -124,6 +139,13 @@ on teston {
...
+  start("A[DsubLoc] = ARepl;");
+  reset();
+  resetA();
+  A[DsubLoc] = ARepl;
+  showA();
+  resetA();
+
   // parallel loop
   start("forall (b,a) in zip(A[Dsub],ARepl) do a = b;");
   reset();
```
Comment (vasslitvinov) [inline test/distributions/vass/testReplicatedDist1.chpl]: How about uncomment the two snippets below and see how they work?

Also, uncommented or not, would you check whether it should be `Dsub` or `DsubLoc` ?

**Comment by bradcray** [comment]
@vasslitvinov: I've added the changes you requested for testReplicatedDist1.chpl.

**Comment by bradcray** [comment]
@vasslitvinov: I was thinking you only reviewed the commit I asked you to, but I just saw that your "approved" seems to have applied to the whole PR.  Did you review the whole thing?  Thanks for clarifying.

**Comment by mppf** [comment]
Did you run GASNet testing?

**Comment by bradcray** [comment]
> Did you run GASNet testing?

Sooo many times now.

**Pull request merged**